boltdb/boltdb_manager.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package boltdb
+
+import (
+	"context"
+	"fmt"
+	"github.com/google/ts-bridge/storage"
+	"os"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/timshannon/bolthold"
+)
+
+// Options holds settings specific to boltdb
+type Options struct {
+	DBPath string
+}
+
+// Manager struct implementing the storage.Manager interface
+type Manager struct {
+	Store *bolthold.Store
+}
+
+// New initializes the Manager struct implementing a generic storage.Manager interface
+func New(_ context.Context, options *Options) *Manager {
+	// If path for a DB is not configured, default to working directory.
+	var boltPath string
+	if options.DBPath == "" {
+		path, err := os.Getwd()
+		if err != nil {
+			log.Fatalf("couldn't get app working directory %v", err)
+		}
+		boltPath = path + "/bolt.db"
+	}
+	store, err := bolthold.Open(boltPath, 0666, nil)
+	if err != nil {
+		log.Fatalf("Unable to Open BoltDB at %v:%v", options.DBPath, err)
+	}
+	log.Debug("Opened BoltDB")
+
+	return &Manager{Store: store}
+}
+
+// NewMetricRecord returns a Datastore-based metric record for a given metric name
+func (d *Manager) NewMetricRecord(_ context.Context, name, query string) (storage.MetricRecord, error) {
+	r := &StoredMetricRecord{Name: name, storage: d}
+	if err := r.load(); err != nil {
+		return nil, err
+	}
+	r.Query = query
+	return r, nil
+}
+
+// CleanupRecords removes obsolete metric records from Datastore.
+func (d *Manager) CleanupRecords(_ context.Context, valid []string) error {
+	// datatype is just an example of the type stored so that the proper bucket and indexes are updated
+	datatype := StoredMetricRecord{}
+	for _, m := range valid {
+		log.Printf("deleting obsolete metric record for %s", m)
+		err := d.Store.Delete(m, datatype)
+		if err != nil {
+			if err == bolthold.ErrNotFound {
+				// This behaviour is copied from the AppEngine storage which ignores
+				// nonexistent records, see datastore/datastore_manager_test.gos
+				log.Printf("record %s does not exist, skipping delete", m)
+			} else {
+				return fmt.Errorf("could not delete metric record %v: %v", m, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// Close closes the BoltDB DB and file
+func (d *Manager) Close() error {
+	err := d.Store.Close()
+	if err != nil {
+		fmt.Errorf("could not close BoltDB store: %v", err)
```

Comment:
Did you mean to not actually return the error?

Perhaps this function should just be `return d.Store.Close()`.

Reply:
That's a fair question to ask. Added a comment to clarify:
```
// explicitly returning an error here since BoltDB file/lock errors can be cryptic without context
```